lib/rspec/core/runner.rb

```diff
@@ -62,7 +62,6 @@ def self.invoke
       #   or the configured failure exit code (1 by default) if specs
       #   failed.
       def self.run(args, err=$stderr, out=$stdout)
-        trap_interrupt
```

Comment:
I don't want to remove this, its important we trap by default during setup in case its expensive, this is what the majority of our users have come to expect, the suggestion above would help with this.

Reply:
What parts of setup are expensive?

```
      def self.run(args, err=$stderr, out=$stdout)
        options = ConfigurationOptions.new(args)

        if options.options[:runner]
          options.options[:runner].call(options, err, out)
        else
          new(options).run(err, out)
        end
      end
```

Are you concerned about `options.options[:runner].call(options, err, out)`?

Reply:
I'm concerned with `setup(err, out)` which loads all the spec files, and thus any requires within them (that often trigger rails booting).

Reply:
If no tests have run, what is the point of handling interrupts during that time?

Reply:
> If no tests have run, what is the point of handling interrupts during that time?

To preserve the behaviour that currently exists.

Reply:
Right, that's a fair point. But what I'm asking is does that behaviour make sense? From what I can tell, this feature is designed to allow breaking out of tests gracefully while still printing some status/test report. But there is no such report when breaking out of the setup phase because no tests have been run.

I would suggest you consider the following:

- The proposed implementation solves the issue, and all specs are passing with no global state.
- It has a well defined behaviour in terms of restoring the signal handling which is absolutely critical if you want to embed this testing framework into something else.
- The counter proposal you suggested involves more global state and doesn't correctly set the signal handling back to the original state.

To me, what concerns me about your proposal is:

- It has more moving parts, some of which are global state.
- It doesn't correctly reset the signal handling after exiting the test runner.

Can we find a solution to those problems?

If you are concerned only about the setup, why don't we wrap that into the interrupt handling?

Reply:
> Right, that's a fair point. But what I'm asking is does that behaviour make sense? From what I can tell, this feature is designed to allow breaking out of tests gracefully while still printing some status/test report. But there is no such report when breaking out of the setup phase because no tests have been run.

Yes it makes sense, it tells people RSpec has not frozen whilst say, Rails is booting, without cancelling a build. 

> I would suggest you consider the following:
>
> * The proposed implementation solves the issue, and all specs are passing with no global state.

The proposed implementation is a breaking change in behaviour and cannot be merged as is.

> * It has a well defined behaviour in terms of restoring the signal handling which is absolutely critical if you want to embed this testing framework into something else.

Whilst we are mindful of how we can be integrated into other tools, our primary usage is as a standalone command line tool, the existing implementation has no signal handling if you instantiate the runner yourself, it is only our short cut run function which does.

> The counter proposal you suggested involves more global state and doesn't correctly set the signal handling back to the original state.
> To me, what concerns me about your proposal is:
>
> It has more moving parts, some of which are global state.
>It doesn't correctly reset the signal handling after exiting the test runner.
>Can we find a solution to those problems?
>
> If you are concerned only about the setup, why don't we wrap that into the interrupt handling?

Signal handlers are global state, there is no getting around that. Your implementation has global state and is hoping that there is no other ruby code in this process or spawned from this process that outlasts your block. I believe that the handler will be passed down to forked children and similarly not test?

I'm open to having both an install method and a remove method, called before and after the class run method. It just needs to replicate the existing behaviour. 

The difficulty is that the handler is global, needs to be installed before setup is run, which might then remove it. This is similar to how our global monkey patching works, for the same reason.

Reply:
e.g. here would call `install_signal_handler!`

Reply:
Yes, you are right, this is modifying global state. However if you wrap your code like this:

```
begin
  previous = Signal.trap(...)
  # user code
ensure
  Signal.trap(..., previous)
end
```

It will be correctly enabled and restored.

Regarding child processes, that is one reason why RSpec should not modify signal handlers at all. So setting it and then setting it back to what it was when it was set the first time, it is not ideal (it could have been changed in the mean time, so setting it back might actually do the wrong thing). It should simply not set it in the first place. The goal of this PR is to provide a way to tell RSpec to NOT call `Signal.trap`.

> Yes it makes sense, it tells people RSpec has not frozen whilst say, Rails is booting, without cancelling a build.

So someone runs RSpec, which loads rails, immediately hits Ctrl-C and it waits for the entire boot process, then exits with an empty summary. I'm not sure why that is preferable over just exiting right away. Can you explain it?

Reply:
> The goal of this PR is to provide a way to tell RSpec to NOT call Signal.trap.

That is not happening. It is not possible to configure RSpec before we need to install the handler. Chicken and egg. 

> So someone runs RSpec, which loads rails, immediately hits Ctrl-C and it waits for the entire boot process, then exits with an empty summary. I'm not sure why that is preferable over just exiting right away. Can you explain it?

If that expensive process is manipulating data it leaves it in a consistent state rather than interrupting half way through.

However why it is preferable or not is personal choice, its not relevant.

What is relevant is that changing the behaviour is from our interpretation a breaking change (which I have mentioned before) in semantic versioning terms, and thus is not possible before RSpec 4 even if there was majority consent for this change.

Reply:
I see.

Well, if users correctly handle interrupts, this should not be an issue. But if that's your decision, I can respect it.

So what about waiting for RSpec 4? I'm fine with that if you think it makes sense.

Because what I want is a way to stop RSpec from registering signal handlers. I'm not interested in the counter proposal you made, it's probably not going to solve my core issues.